src/fetch-info.js

```diff
@@ -0,0 +1,264 @@
+/**
+ * Module that exports a function that takes an array of specifications objects
+ * that each have at least a "url" and a "name" property. The function returns
+ * an object indexed by specification "name" with additional information
+ * about the specification fetched from the W3C API, Specref, or from the spec
+ * itself. Object returned for each specification contains the following
+ * properties:
+ *
+ * - "edUrl": the URL of the Editor's Draft of the specification. Always set.
+ * - "trUrl": the URL of the latest version of the specification published to
+ * /TR/. Only set when the specification has been published there.
+ * - "title": the title of the specification. Always set.
+ * - "source": one of "w3c", "specref", "spec", depending on how the information
+ * was determined.
+ *
+ * The function throws when something goes wrong, e.g. if the given spec object
+ * describes a /TR/ specification but the specification has actually not been
+ * published to /TR/, if the specification cannot be fetched, or if no W3C API
+ * key was specified for a /TR/ URL.
+ *
+ * The function will start by querying the W3C API, using the given "name"
+ * properties. For specifications where this fails, the function will query
+ * SpecRef, using the given "name" as well. If that too fails, the function
+ * assumes that the given "url" is the URL of the Editor's Draft, and will fetch
+ * it to determine the title.
+ *
+ * The function needs an API key to fetch the W3C API, which can be passed
+ * within an "options" object with a "w3cApiKey" property.
+ *
+ * If the function needs to retrieve the spec itself, note that it will parse
+ * the HTTP response body as a string, applying regular expressions to extract
+ * the title. It will not parse it as HTML in particular. This means that the
+ * function will fail if the title cannot easily be extracted for some reason.
+ *
+ * Note: the function operates on a list of specs and not only on one spec to
+ * bundle requests to Specref.
+ */
+
+const https = require("https");
+
+
+async function fetchInfoFromW3CApi(specs, options) {
+  // Cannot query the W3C API if API key was not given
+  if (!options || !options.w3cApiKey) {
+    return [];
+  }
+  options.headers = options.headers || {};
+  options.headers.Authorization = `W3C-API apikey="${options.w3cApiKey}"`;
+
+  const info = await Promise.all(specs.map(async spec => {
+    // Skip specs when the known URL is not a /TR/ URL, because there may still
+    // be a spec with the same name published to /TR/ but that is probably an
+    // outdated version (e.g. WHATWG specs such as DOM or Fullscreen, or CSS
+    // drafts published a long long time ago)    
+    if (!spec.url.match(/^https?:\/\/(?:www\.)?w3\.org\/TR\/([^\/]+)\/$/)) {
+      return;
+    }
+
+    const url = `https://api.w3.org/specifications/${spec.name}`;
+    return new Promise((resolve, reject) => {
+      const request = https.get(url, options, res => {
```

Comment:
any reason you're not using a higher level library (e.g. requests) for HTTP fetches?

Reply:
Yes, I'm trying to introduce dependencies only when needed. We'll probably need to change that later on but network requests seem manageable with core node.js modules for now.